spring-session/src/main/java/org/springframework/session/jdbc/JdbcOperationsSessionRepository.java

```diff
@@ -0,0 +1,439 @@
+				public void setValues(PreparedStatement ps) throws SQLException {
+					ps.setLong(1, session.getLastAccessedTime());
+					ps.setString(2, session.getPrincipalName());
+					JdbcOperationsSessionRepository.this.lobHandler.getLobCreator()
+							.setBlobAsBytes(ps, 3, sessionBytes);
+					ps.setString(4, session.getId());
+				}
+
+			});
+		}
+	}
+
+	@Override
+	public JdbcSession getSession(String id) {
+		MapSession session = null;
+		try {
+			session = this.jdbcOperations.queryForObject(getQuery(Query.GET_SESSION),
+					new Object[] { id }, this.mapper);
+		}
+		catch (EmptyResultDataAccessException ignored) {
+		}
+
+		if (session != null) {
+			if (session.isExpired()) {
+				delete(id, new SessionExpiredEvent(this, id));
+			}
+			else {
+				return new JdbcSession(session);
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public void delete(String id) {
+		delete(id, new SessionDeletedEvent(this, id));
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public Map<String, JdbcSession> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		if (!PRINCIPAL_NAME_INDEX_NAME.equals(indexName)) {
+			return Collections.emptyMap();
+		}
+
+		List<MapSession> sessions = this.jdbcOperations.query(
+				getQuery(Query.LIST_SESSIONS_BY_PRINCIPAL_NAME), new Object[] { indexValue }, this.mapper);
+
+		Map<String, JdbcSession> sessionMap = new HashMap<String, JdbcSession>(sessions.size());
+
+		for (MapSession session : sessions) {
+			sessionMap.put(session.getId(), new JdbcSession(session));
+		}
+
+		return sessionMap;
+	}
+
+	@Scheduled(cron = "0 * * * * *")
+	public void cleanupExpiredSessions() {
+		long now = System.currentTimeMillis();
+		long roundedNow = roundDownMinute(now);
+
+		if (logger.isDebugEnabled()) {
+			logger.debug("Cleaning up sessions expiring at " + new Date(roundedNow));
+		}
+
+		List<String> sessionIds = this.jdbcOperations.queryForList(
+				getQuery(Query.LIST_SESSIONS_IDS_BY_LAST_ACCESS_TIME), new Object[] { roundedNow }, String.class);
+
+		for (String sessionId : sessionIds) {
+			delete(sessionId, new SessionExpiredEvent(this, sessionId));
+		}
+	}
+
+	/**
+	 * Replace patterns in the input to produce a valid SQL query. This implementation lazily initializes a
+	 * simple map-based cache, only replacing the table prefix on the first access to a named query. Further
+	 * accesses will be resolved from the cache.
+	 *
+	 * @param base the SQL query to be transformed
+	 * @return a transformed query with replacements
+	 */
+	protected String getQuery(Query base) {
+		String query = this.queryCache.get(base);
+
+		if (query == null) {
+			query = StringUtils.replace(base.getSql(), "%PREFIX%", this.tablePrefix);
+			this.queryCache.put(base, query);
+		}
+
+		return query;
+	}
+
+	public void delete(String id, SessionDestroyedEvent event) {
+		this.jdbcOperations.update(getQuery(Query.DELETE_SESSION), id);
+		publishEvent(event);
+	}
+
+	private void publishEvent(ApplicationEvent event) {
```

Comment:
Unfortunately, this will not publish the ApplicationEvent for every node in the cluster. For this reason, I think we should just not publish the event and establish this as a limitation of the datastore.

This ties in with why we can simply perform a delete rather than looking up the session id first.